Std/Data/String/Basic.lean

```diff
@@ -1,17 +1,83 @@
 /-
 Copyright (c) 2022 Jannis Limperg. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Jannis Limperg
+Authors: Jannis Limperg, James Gallicchio, F. G. Dorais
 -/
 
 import Std.Data.Char
 import Std.Data.Nat.Lemmas
+import Std.Data.Array.Match
 
 instance : Coe String Substring := ⟨String.toSubstring⟩
 
-protected theorem String.Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
+namespace String
+
+protected theorem Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
 | _, _, hlt, rfl => Nat.not_lt_zero _ hlt
 
+/-- Knuth-Morris-Pratt matcher type
+
+This type is used to keep data for running the Knuth-Morris-Pratt (KMP) string matching algorithm.
+KMP is a linear time algorithm to locate all substrings of a string that match a given pattern.
+Generating the algorithm data is also linear in the length of the pattern but the data can be
+re-used to match the same pattern over different strings.
+
+The KMP data for a pattern string can be generated using `Matcher.ofString`. Then `Matcher.find?`
+and `Matcher.findAll` can be used to run the algorithm on an input string.
+```
+def m := Matcher.ofString "abba"
+
+#eval Option.isSome <| m.find? "AbbabbA" -- false
+#eval Option.isSome <| m.find? "aabbaa" -- true
+
+#eval Array.size <| m.findAll "abbabba" -- 2
+#eval Array.size <| m.findAll "abbabbabba" -- 3
+```
+-/
+abbrev Matcher := Array.Matcher Char
+
+/-- Make KMP matcher from pattern substring -/
+@[inline] def Matcher.ofSubstring (pattern : Substring) : Matcher :=
+  Array.Matcher.ofStream pattern
+
+/-- Make KMP matcher from pattern string -/
+@[inline] def Matcher.ofString (pattern : String) : Matcher :=
+  Matcher.ofSubstring pattern
+
+/-- The string pattern for the matcher -/
+def Matcher.pattern (m : Matcher) : String :=
+  m.table.foldl (fun s (c,_) => s.push c) ""
+
+/-- The byte size of the string pattern for the matcher -/
+def Matcher.patternSize (m : Matcher) : Nat :=
+  String.Pos.byteIdx <| m.table.foldl (fun n (c,_) => n + c) 0
+
+/-- Find all substrings of `s` matching `m.pattern`. -/
+partial def Matcher.findAll (m : Matcher) (s : Substring) : Array Substring :=
+  loop s m #[]
+where
+  /-- Byte size of match pattern -/
+  size := m.patternSize
+  /-- Accumulator loop for `String.Matcher.findAll` -/
+  loop (s : Substring) (m : Matcher) (occs : Array Substring) : Array Substring :=
+    match m.next? s with
+    | none => occs
+    | some (s, m) =>
+      loop s m <| occs.push { s with
+        startPos := ⟨s.startPos.byteIdx - size⟩
+        stopPos := s.startPos }
+
```

Comment:
Not a particularly serious suggestion, but you could produce a lazy list of matches here. :-)

Reply:
Good question! This is a definite possibility using `MLList`. I'll think about it. If it turns out to be straightforward I will include it here but a separate PR might be a better venue.

Reply:
I thought about it for a bit. I like the idea but I think this would belong in a later PR. There seems to be a few choices available (`MLList` and `Stream` are the top two). Also, `Array.Match.next?` is already there as a low-level infrastructure to generate matches lazily. Since you said this was not particularly serious, I will politely pass.